nCompiler/inst/include/nCompiler/repClass.h

```diff
@@ -0,0 +1,106 @@
+#ifndef _NCOMPILER_REP_CLASS
+#define _NCOMPILER_REP_CLASS
+
+template <typename DerivedOut>
+struct rep_impl {
+
+  typedef typename DerivedOut::Scalar ScalarType;
+  typedef typename DerivedOut::Index IndexType;
+
+  static Eigen::Tensor<ScalarType, 1> maybe_reshape(DerivedOut x) {
+    Eigen::Tensor<ScalarType, 1> reshaped;
+    std::array<IndexType, 1> one_dim{{static_cast<IndexType>(x.size())}};
+    // from Tensor docs: "This operation does not move any data in the input
+    // tensor, so the resulting contents of a reshaped Tensor depend on the
+    // data layout of the original Tensor."
+    reshaped = x.reshape(one_dim);
+    return(reshaped);
+  }
+
+  static Eigen::Tensor<ScalarType, 1> repTimes(DerivedOut x, int times) {
+    std::array<IndexType, 1> bcast{{times}};
+    // auto here hopefully avoids materializing x too early?
+    auto reshaped = maybe_reshape(x);
+    return(reshaped.broadcast(bcast));
+  }
+
+  static Eigen::Tensor<ScalarType, 1> repTimesEval(DerivedOut x, int times) {
+    std::array<IndexType, 1> bcast{{times}};
+    auto reshaped = maybe_reshape(x);
+    return(reshaped.eval().broadcast(bcast));
+  }
+
+  static Eigen::Tensor<ScalarType, 1> repLen(DerivedOut x, int length_out) {
+    std::array<IndexType, 1> offset{{0}};
+    std::array<IndexType, 1> extent{{length_out}};
+    auto reshaped = maybe_reshape(x);
+    if (length_out < x.size()) {
+      return(reshaped.slice(offset, extent));
+    } else {
+      int times = ceil(static_cast<double>(length_out) / x.size());
+      std::array<IndexType, 1> bcast{{times}};
+      return(reshaped.broadcast(bcast).slice(offset, extent));
+    }
+  }
+
+  static Eigen::Tensor<ScalarType, 1> repLenEval(DerivedOut x, int length_out) {
+    std::array<IndexType, 1> offset{{0}};
+    std::array<IndexType, 1> extent{{length_out}};
+    auto reshaped = maybe_reshape(x);
+    if (length_out < x.size()) {
+      return(reshaped.slice(offset, extent));
+    } else {
+      int times = ceil(static_cast<double>(length_out) / x.size());
+      std::array<IndexType, 1> bcast{{times}};
+      return(reshaped.eval().broadcast(bcast).slice(offset, extent));
+    }
+  }
+};
+
+// TODO: support arbitrary NumDimensions
+// TODO: implement repEach
+// repTimes
+#define repTimes1d rep_impl<Eigen1d>::repTimes
+#define repTimes2d rep_impl<Eigen2d>::repTimes
+#define repTimes3d rep_impl<Eigen3d>::repTimes
+#define repTimes1i rep_impl<Eigen1i>::repTimes
+#define repTimes2i rep_impl<Eigen2i>::repTimes
+#define repTimes3i rep_impl<Eigen3i>::repTimes
+#define repTimes1b rep_impl<Eigen1b>::repTimes
+#define repTimes2b rep_impl<Eigen2b>::repTimes
+#define repTimes3b rep_impl<Eigen3b>::repTimes
+
+// repTimesEval
+#define repTimesEval1d rep_impl<Eigen1d>::repTimesEval
```

Comment:
Could insert the `.eval()` into the AST directly before passing the arg to `repTimes`.